src/coreclr/src/vm/jithelpers.cpp

```diff
@@ -3001,84 +3003,64 @@ HCIMPL3(VOID, JIT_Unbox_Nullable, void * destPtr, CORINFO_CLASS_HANDLE type, Obj
+
+    if (pMT1->GetVerifierCorElementType() == pMT2->GetVerifierCorElementType())
+    {
+        // we allow enums and their primitive type to be interchangable
+        result = objRef->GetData();
+    }
+    else if (pMT1->IsEquivalentTo(pMT2))
     {
         // the structures are equivalent
         result = objRef->GetData();
     }
     else
     {
-        COMPlusThrowInvalidCastException(&objRef, TypeHandle(type));
+        COMPlusThrowInvalidCastException(&objRef, TypeHandle(pMT1));
     }
+
+    HELPER_METHOD_POLL(); 
     HELPER_METHOD_FRAME_END();
 
     return result;
 }
 HCIMPLEND
 
 /*************************************************************/
-/* the uncommon case for the helper below (allowing enums to be unboxed
-   as their underlying type */
-LPVOID __fastcall JIT_Unbox_Helper(CORINFO_CLASS_HANDLE type, Object* obj)
+/* Unbox helper that handles enums */
+HCIMPL2(LPVOID, Unbox_Helper, CORINFO_CLASS_HANDLE type, Object* obj)
 {
     FCALL_CONTRACT;
 
     TypeHandle typeHnd(type);
+    // boxable types have method tables
+    _ASSERTE(!typeHnd.IsTypeDesc());
 
-    CorElementType type1 = typeHnd.GetInternalCorElementType();
-
-        // we allow enums and their primtive type to be interchangable
+    MethodTable* pMT1 = typeHnd.AsMethodTable();
+    // must be a value type
+    _ASSERTE(pMT1->IsValueType());
 
     MethodTable* pMT2 = obj->GetMethodTable();
-    CorElementType type2 = pMT2->GetInternalCorElementType();
-    if (type1 == type2)
-    {
-        MethodTable* pMT1 = typeHnd.AsMethodTable();
-        if (pMT1 && (pMT1->IsEnum() || pMT1->IsTruePrimitive()) &&
-            (pMT2->IsEnum() || pMT2->IsTruePrimitive()))
-        {
-            _ASSERTE(CorTypeInfo::IsPrimitiveType_NoThrow(type1));
-            return(obj->GetData());
-        }
-    }
-
-    // Even less common cases (type equivalence) go to a framed helper.
-    ENDFORBIDGC();
-    return HCCALL2(JIT_Unbox_Helper_Framed, type, obj);
-}
-
-/*************************************************************/
-HCIMPL2(LPVOID, JIT_Unbox, CORINFO_CLASS_HANDLE type, Object* obj)
-{
-    FCALL_CONTRACT;
 
-    TypeHandle typeHnd(type);
-    VALIDATEOBJECT(obj);
-    _ASSERTE(!typeHnd.IsTypeDesc());   // boxable types have method tables
-
-        // This has been tuned so that branch predictions are good
-        // (fall through for forward branches) for the common case
-    if (obj != NULL) {
-        if (obj->GetMethodTable() == typeHnd.AsMethodTable())
-            return(obj->GetData());
-        else {
-                // Stuff the uncommon case into a helper so that
-                // its register needs don't cause spills that effect
-                // the common case above.
-            return JIT_Unbox_Helper(type, obj);
-        }
+    // we allow enums and their primitive type to be interchangable.
+    // if suspension is requested, defer to the framed helper.
+    if (pMT1->GetVerifierCorElementType() == pMT2->GetVerifierCorElementType() &&
```

Comment:
This also needs to check that the CorElementType is primitive (ie `!= ELEMENTTYPE_VALUETYPE`?)

Reply:
It looks like the fastest check is
```cs
    if (pMT1->GetInternalCorElementType() == pMT2->GetInternalCorElementType() &&
        (pMT1->IsEnum() || pMT1->IsTruePrimitive()) &&
        (pMT2->IsEnum() || pMT2->IsTruePrimitive()))
    {
```

Reply:
Why not this?
```
CorElementType corElementType = pMT1->GetInternalCorElementType();
if (corElementType != ELEMENT_TYPE_VALUETYPE &&  corElementType == pMT2->GetInternalCorElementType())
{
```

Reply:
It is yet a bit faster. :-)